File: integration/examples/python/smoke.py
"""Smoke test: S3 put+get and a DynamoDB round-trip against aws-local-lab.

Exits 0 on success, non-zero on any failure. Safe to run repeatedly.
"""
from __future__ import annotations

import sys
import time
import uuid

from botocore.exceptions import ClientError

from client import client

BUCKET = "integration-smoke-py"
TABLE = "integration-smoke-py"


def wait_for_edge(timeout: int = 60) -> None:
    s3 = client("s3")
    deadline = time.time() + timeout
    while True:
        try:
            s3.list_buckets()
            return
        except Exception as exc:  # noqa: BLE001
            if time.time() > deadline:
                raise
            print(f"waiting for lab edge... ({exc})")
            time.sleep(2)


def ensure_bucket(s3) -> None:
    try:
        s3.head_bucket(Bucket=BUCKET)
    except ClientError:
        s3.create_bucket(Bucket=BUCKET)


def ensure_table(ddb) -> None:
    existing = ddb.list_tables()["TableNames"]
    if TABLE not in existing:
        ddb.create_table(
            TableName=TABLE,
            KeySchema=[{"AttributeName": "id", "KeyType": "HASH"}],
            AttributeDefinitions=[{"AttributeName": "id", "AttributeType": "S"}],
            BillingMode="PAY_PER_REQUEST",
        )
        ddb.get_waiter("table_exists").wait(TableName=TABLE)


def main() -> int:
    wait_for_edge()
    s3 = client("s3")
    ddb = client("dynamodb")

    ensure_bucket(s3)
    ensure_table(ddb)

    key = f"obj-{uuid.uuid4()}"
    body = f"hello-{uuid.uuid4()}".encode()
    s3.put_object(Bucket=BUCKET, Key=key, Body=body)
    got = s3.get_object(Bucket=BUCKET, Key=key)["Body"].read()
    assert got == body, f"S3 body mismatch: {got!r} != {body!r}"
    print(f"[python] S3 put/get OK  s3://{BUCKET}/{key} ({len(body)} bytes)")

    item_id = str(uuid.uuid4())
    ddb.put_item(TableName=TABLE, Item={"id": {"S": item_id}, "v": {"S": "round-trip"}})
    read = ddb.get_item(TableName=TABLE, Key={"id": {"S": item_id}})["Item"]
    assert read["v"]["S"] == "round-trip", f"DynamoDB mismatch: {read!r}"
    print(f"[python] DynamoDB round-trip OK  {TABLE}/{item_id}")

    print("[python] SMOKE PASS")
    return 0


if __name__ == "__main__":
    try:
        sys.exit(main())
    except Exception as exc:  # noqa: BLE001
        print(f"[python] SMOKE FAIL: {exc}", file=sys.stderr)
        sys.exit(1)

File: integration/examples/node/smoke.mjs
// Smoke test: S3 put+get and a DynamoDB round-trip against aws-local-lab.
// Exits 0 on success, non-zero on failure. Safe to run repeatedly.
import { randomUUID } from "node:crypto";
import {
  CreateBucketCommand,
  GetObjectCommand,
  HeadBucketCommand,
  PutObjectCommand,
} from "@aws-sdk/client-s3";
import {
  CreateTableCommand,
  DescribeTableCommand,
  GetItemCommand,
  ListTablesCommand,
  PutItemCommand,
} from "@aws-sdk/client-dynamodb";
import { s3 as makeS3, dynamodb as makeDdb } from "./client.mjs";

const BUCKET = "integration-smoke-node";
const TABLE = "integration-smoke-node";

const sleep = (ms) => new Promise((r) => setTimeout(r, ms));

async function waitForEdge(ddb, timeoutMs = 60000) {
  const deadline = Date.now() + timeoutMs;
  for (;;) {
    try {
      await ddb.send(new ListTablesCommand({}));
      return;
    } catch (err) {
      if (Date.now() > deadline) throw err;
      console.log(`waiting for lab edge... (${err.name})`);
      await sleep(2000);
    }
  }
}

async function ensureBucket(s3) {
  try {
    await s3.send(new HeadBucketCommand({ Bucket: BUCKET }));
  } catch {
    await s3.send(new CreateBucketCommand({ Bucket: BUCKET }));
  }
}

async function ensureTable(ddb) {
  const { TableNames } = await ddb.send(new ListTablesCommand({}));
  if (TableNames.includes(TABLE)) return;
  await ddb.send(
    new CreateTableCommand({
      TableName: TABLE,
      KeySchema: [{ AttributeName: "id", KeyType: "HASH" }],
      AttributeDefinitions: [{ AttributeName: "id", AttributeType: "S" }],
      BillingMode: "PAY_PER_REQUEST",
    }),
  );
  for (let i = 0; i < 30; i++) {
    const { Table } = await ddb.send(new DescribeTableCommand({ TableName: TABLE }));
    if (Table.TableStatus === "ACTIVE") return;
    await sleep(1000);
  }
}

async function streamToString(stream) {
  const chunks = [];
  for await (const chunk of stream) chunks.push(chunk);
  return Buffer.concat(chunks).toString("utf8");
}

async function main() {
  const s3 = makeS3();
  const ddb = makeDdb();

  await waitForEdge(ddb);
  await ensureBucket(s3);
  await ensureTable(ddb);

  const key = `obj-${randomUUID()}`;
  const body = `hello-${randomUUID()}`;
  await s3.send(new PutObjectCommand({ Bucket: BUCKET, Key: key, Body: body }));
  const got = await s3.send(new GetObjectCommand({ Bucket: BUCKET, Key: key }));
  const text = await streamToString(got.Body);
  if (text !== body) throw new Error(`S3 body mismatch: ${text} != ${body}`);
  console.log(`[node] S3 put/get OK  s3://${BUCKET}/${key} (${body.length} bytes)`);

  const id = randomUUID();
  await ddb.send(
    new PutItemCommand({ TableName: TABLE, Item: { id: { S: id }, v: { S: "round-trip" } } }),
  );
  const read = await ddb.send(
    new GetItemCommand({ TableName: TABLE, Key: { id: { S: id } } }),
  );
  if (read.Item?.v?.S !== "round-trip") throw new Error(`DynamoDB mismatch: ${JSON.stringify(read.Item)}`);
  console.log(`[node] DynamoDB round-trip OK  ${TABLE}/${id}`);

  console.log("[node] SMOKE PASS");
}

main().catch((err) => {
  console.error(`[node] SMOKE FAIL: ${err.message}`);
  process.exit(1);
});
